File: frontend/src/types/task.ts
export type AgentProgress = {
  status: "pending" | "running" | "completed" | "error" | string;
  duration_ms?: number | null;
  error?: string;
  input_summary?: string;
  output_summary?: string;
};

export type TaskStatus = {
  task_id: number;
  status: string;
  progress: Record<string, AgentProgress | string>;
  error_message: string;
  cancel_requested: boolean;
  run_attempt: number;
  queued_at?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
};

export type Task = {
  id: number;
  name: string;
  domain: string;
  platforms: string[];
  board: string;
  match_name: string;
  home_team: string;
  away_team: string;
  match_stage: string;
  match_date: string;
  thread_urls: string[];
  news_urls: string[];
  news_context: string;
  keywords: string[];
  semantic_query: string;
  time_range: Record<string, unknown>;
  max_comments: number;
  similarity_threshold: number;
  language: string;
  sentiment_focus: string;
  enable_llm: boolean;
  llm_mode: "mock" | "configured" | string;
  enable_image_analysis: boolean;
  max_image_comments: number;
  data_source: string;
  source_path?: string | null;
  upload_id?: string | null;
  field_mapping: Record<string, string>;
  status: string;
  progress: Record<string, AgentProgress | string>;
  error_message: string;
  cancel_requested: boolean;
  run_attempt: number;
  queued_at?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  updated_at: string;
};

export type CommentItem = {
  id: number;
  raw_comment_id: string;
  platform: string;
  content: string;
  cleaned_content: string;
  language: string;
  quality_score: number;
  is_duplicate: boolean;
  like_count: number;
  sentiment_label?: string | null;
  sentiment_score?: number | null;
  painpoint?: string | null;
  positive_attribution?: string | null;
  representative_reason?: string | null;
  cluster_id?: number | null;
  image_urls: string[];
  image_analysis: {
    status?: string;
    model?: string;
    summary?: string;
    ocr_text?: string;
    entities?: string[];
    relevance?: string;
    sentiment_cue?: string;
    error?: string;
  };
};

export type Cluster = {
  cluster_id: number;
  cluster_name: string;
  cluster_size: number;
  cluster_ratio: number;
  top_keywords: string[];
  sentiment_distribution: Record<string, number>;
  method: string;
  is_noise: boolean;
  representative_comments: string[];
};

export type DataQuality = {
  raw_count: number;
  clean_count: number;
  dedup_count: number;
  duplicate_ratio: number;
  noise_ratio: number;
  language_distribution: Record<string, number>;
  sample_confidence_level: "high" | "medium" | "low" | string;
  warning: string;
};

export type WordCloudItem = {
  word: string;
  weight: number;
};

export type WordClouds = {
  all_words: WordCloudItem[];
  positive_words: WordCloudItem[];
  negative_words: WordCloudItem[];
  explanations: Record<string, string>;
};

export type Insight = {
  summary: string;
  positive_insights: string[];
  negative_insights: string[];
  key_viewpoints: string[];
  controversies: string[];
  news_context_summary: string;
  context_alignment: string[];
  fact_opinion_gaps: string[];
  risks: string[];
  context_media: ContextMedia[];
};

export type ContextMedia = {
  source_id: string;
  source_kind: "hupu_thread" | "news" | string;
  title: string;
  text_context: string;
  source_url: string;
  image_urls: string[];
  authority_level: "official" | "official_reference" | "editorial" | "community_data" | string;
  information_score: number;
  selection_reasons: string[];
  status: string;
  model?: string;
  summary?: string;
  ocr_text?: string;
  entities?: string[];
  data_points?: string[];
  relevance?: string;
  information_value?: string;
  confidence?: string;
  included_in_summary: boolean;
  exclusion_reason?: string;
  error?: string;
};

export type ClassificationRow = {
  category: string;
  count: number;
  ratio: number;
  examples: string[];
};

export type UploadedDataset = {
  id: string;
  original_name: string;
  source_kind: "csv" | "json" | "mediacrawler" | string;
  file_format: string;
  size_bytes: number;
  row_count: number;
  columns: string[];
  preview_rows: Record<string, unknown>[];
  field_mapping: Record<string, string>;
  validation_errors: string[];
  status: "ready" | "needs_mapping" | string;
  created_at: string;
};

export type StrategyEvidence = {
  comment_id: string;
  content: string;
  source_url?: string;
  like_count?: number;
};

export type StrategyCard = {
  id: number;
  task_id: number;
  title: string;
  card_type: string;
  evidence_comment_ids: string[];
  evidence_comments: StrategyEvidence[];
  evidence_count: number;
  sample_size: number;
  affected_ratio: number;
  confidence: string;
  confidence_reason: string;
  suggested_actions: string[];
  expected_impact: string;
  ab_test_design: Record<string, unknown>;
  created_at: string;
};

export type ABTestDraft = {
  id: number;
  task_id: number;
  strategy_card_id: number;
  name: string;
  hypothesis: string;
  control: string;
  variant: string;
  primary_metric: string;
  guardrail_metrics: string[];
  sample_size_note: string;
  status: string;
  created_at: string;
};

File: backend/app/schemas/task.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class TaskCreate(BaseModel):
    """Payload for creating an analysis task."""

    name: str = Field(default="虎扑赛事评论分析", min_length=1)
    domain: str = Field(default="basketball", pattern="^(basketball|football)$")
    platforms: list[str] = Field(default_factory=lambda: ["hupu"])
    board: str = "nba"
    match_name: str = ""
    home_team: str = ""
    away_team: str = ""
    match_stage: str = ""
    match_date: str = ""
    thread_urls: list[str] = Field(default_factory=list)
    news_urls: list[str] = Field(default_factory=list)
    news_context: str = ""
    keywords: list[str] = Field(default_factory=list)
    semantic_query: str = "分析虎扑网友对本场比赛的情绪、争议焦点和主要观点"
    time_range: dict[str, Any] = Field(default_factory=dict)
    max_comments: int = Field(default=120, ge=1, le=5000)
    similarity_threshold: float = Field(default=0.86, ge=0.0, le=1.0)
    language: str = "zh"
    sentiment_focus: str = "all"
    enable_llm: bool = True
    enable_image_analysis: bool = True
    max_image_comments: int = Field(default=6, ge=0, le=20)
    data_source: str = Field(default="mock", pattern="^(mock|hupu_public|csv|json)$")
    source_path: str | None = None


class TaskOut(BaseModel):
    """Task response model."""

    id: int
    name: str
    domain: str
    platforms: list[str]
    board: str
    match_name: str
    home_team: str
    away_team: str
    match_stage: str
    match_date: str
    thread_urls: list[str]
    news_urls: list[str]
    news_context: str
    keywords: list[str]
    semantic_query: str
    time_range: dict[str, Any]
    max_comments: int
    similarity_threshold: float
    language: str
    sentiment_focus: str
    enable_llm: bool
    enable_image_analysis: bool
    max_image_comments: int
    data_source: str
    source_path: str | None
    status: str
    progress: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskStatusOut(BaseModel):
    """Current task status and stage progress."""

    task_id: int
    status: str
    progress: dict[str, Any]


class TaskRunResponse(BaseModel):
    """Response returned after a task run is triggered."""

    task_id: int
    status: str
    summary: dict[str, Any]


class CommentOut(BaseModel):
    """Comment API response after cleaning and annotation."""

    id: int
    raw_comment_id: str
    platform: str
    content: str
    cleaned_content: str
    language: str
    quality_score: float
    is_duplicate: bool
    like_count: int
    sentiment_label: str | None = None
    sentiment_score: float | None = None
    painpoint: str | None = None
    positive_attribution: str | None = None
    representative_reason: str | None = None
    image_urls: list[str] = Field(default_factory=list)
    image_analysis: dict[str, Any] = Field(default_factory=dict)


class SentimentOut(BaseModel):
    """Sentiment distribution response."""

    distribution: dict[str, int]
    average_score: float


class PainPointOut(BaseModel):
    """Pain point top-N response."""

    category: str
    count: int
    ratio: float
    examples: list[str]


class PositiveAttributionOut(BaseModel):
    """Positive attribution top-N response."""

    category: str
    count: int
    ratio: float
    examples: list[str]


class ClusterOut(BaseModel):
    """Cluster summary response."""

    cluster_id: int
    cluster_name: str
    cluster_size: int
    cluster_ratio: float
    top_keywords: list[str]
    sentiment_distribution: dict[str, int]
    method: str = "rules"
    is_noise: bool = False
    representative_comments: list[str] = Field(default_factory=list)


class DataQualityOut(BaseModel):
    """Data quality report response."""

    raw_count: int
    clean_count: int
    dedup_count: int
    duplicate_ratio: float
    noise_ratio: float
    language_distribution: dict[str, int]
    sample_confidence_level: str
    warning: str = ""


class WordCloudItem(BaseModel):
    """Weighted word item for word cloud rendering."""

    word: str
    weight: float


class WordCloudsOut(BaseModel):
    """All, positive and negative TF-IDF word clouds."""

    all_words: list[WordCloudItem] = Field(default_factory=list)
    positive_words: list[WordCloudItem] = Field(default_factory=list)
    negative_words: list[WordCloudItem] = Field(default_factory=list)
    explanations: dict[str, str] = Field(default_factory=dict)


class InsightOut(BaseModel):
    """Structured insight report response."""

    summary: str
    positive_insights: list[str]
    negative_insights: list[str]
    key_viewpoints: list[str]
    controversies: list[str]
    news_context_summary: str
    context_alignment: list[str]
    fact_opinion_gaps: list[str]
    risks: list[str]
    context_media: list[dict[str, Any]] = Field(default_factory=list)
